### PR TITLE
Fix literal rounding in codegen functions

### PR DESCRIPTION
**Pull request by kahmed10**
Fixes the failing test case in #1815. Added a test that would otherwise fail without the change.

### CODE REVIEW

**Review by TedThemistokleous — APPROVED** [review]

File: test/gpu/codegen_literal.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include "migraphx/op/mul.hpp"
+#include <iostream>
+#include <vector>
+#include <migraphx/operators.hpp>
```
Comment (pfultz2) [inline test/gpu/codegen_literal.cpp:27]: Dont include this header.

File: test/gpu/codegen_literal.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+#include <migraphx/instruction.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/register_target.hpp>
+#include <migraphx/verify.hpp>
+#include <migraphx/pass_manager.hpp>
```
Comment (pfultz2) [inline test/gpu/codegen_literal.cpp:32]: Is this needed? I dont see it using passes.

File: test/gpu/codegen_literal.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+#include <test.hpp>
+#include <migraphx/make_op.hpp>
+
+// This test ensures that the codegen path doesn't round up literals,
+// otherwise there are accuracy differences compared to ref.
+// The values being passed in are 0.5 * (1/0.00787402),
+// and after rounding must equal 63, not 64.
+TEST_CASE(mul_literal_round_test)
+{
+    auto run_prog = [](migraphx::program p,
```
Comment (pfultz2) [inline test/gpu/codegen_literal.cpp:42]: Make this a free function.

File: test/gpu/codegen_literal.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            {
+                m[x.first] = t.allocate(x.second);
+            }
+        }
+
+        auto result = t.copy_from(p.eval(m).back());
+        result.visit([&](auto v) { res.assign(v.begin(), v.end()); });
+    };
+
+    auto create_program = [&] {
```
Comment (pfultz2) [inline test/gpu/codegen_literal.cpp:64]: Inline this.

File: test/gpu/codegen_literal.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include "migraphx/op/mul.hpp"
```
Comment (pfultz2) [inline test/gpu/codegen_literal.cpp:24]: Dont include this header.

File: test/gpu/codegen_literal.cpp (hunk 2 above)
Comment (kahmed10) [reply test/gpu/codegen_literal.cpp:32]: Probably not, going to clean up and remove unneeded headers

**Comment by pfultz2** [comment]
Is it possible to add a `to_string` unit test as well? 

```cpp
TEST_CASE(literal_to_string_float_precision)
{
    migraphx::literal x{1 / 0.00787402f};
    EXPECT(x.to_string() == "126.99993142003703");
}
```

This can probably go in `test/literal_test.cpp`.

**Comment by kahmed10** [comment]
> ```c++
> migraphx::literal x{1 / 0.00787402f};
>     EXPECT(x.to_string() == "126.99993142003703");
> ```

Seems like this test fails because the string value doesn't match directly. I think I will rewrite it to prevent rounding.

**Pull request merged**